### PR TITLE
Add trio111, passing variables from context managers to nurseries opened outside them

### PR DESCRIPTION
**Pull request by jakkdl**
A couple test cases I'm unsure how they should be handled, so make sure to check the test file thoroughly.

I'm also slightly confused about your definition of inner/outer - you say "should be outermost" but https://github.com/python-trio/trio/issues/671#issuecomment-533978268 shows
```python
async with trio.open_nursery():
  async with trio.open_process():
    ...
```
being troublesome, which I'd call the nursery being the outer scope - and as such nurseries should be "innermost".

But easy to reverse logic / error message as needed.

### CODE REVIEW

**Comment by Zac-HD** [comment]
Via https://github.com/python-trio/trio/issues/671#issuecomment-533989210:

> I guess the general case of detecting when something is closed while in use would require some super sophisticated data-flow analysis, but this specific pattern of `async with ... as x` around a call to `start_soon(..., x)` seems like it could be simple enough for a linter to detect + a common enough error to be a useful lint.
>

> Though we might want to special-case `async with open_nursery() as nursery: nursery.start_soon(..., nursery)`, since that's not an error. More generally, I guess the point is that we need to also have some heuristic to rule out cases where the `async with` surrounds the nursery.

Yeah, you're entirely right; the nursery needs to be innermost not outermost.

Important to note though that (to reduce false-alarms) we only want to warn if we have something from a context manager *passed to* the task(s) being started, and the context will exit before the nursery exits.

We'll also want to check for `nursery.start()` as well as `.start_soon()`.

**Comment by jakkdl** [comment]
Ah, so:
* If a nursery is opened and saved in variable `X`
* and inside that nursery a trio (or all?) async context manager is opened and saved in variable `Y`
* and inside that context manager there's a call `X.start[_soon]([...], Y, [...])`

**Comment by Zac-HD** [comment]
That's right! I'd check for any sync-or-async context manager to start.

**Comment by jakkdl** [comment]
Haven't made a second pass through the code to clean up and add comments and stuff, but pushing a dirty version so you can check if the tests fit the specification.

The error message is also ... a WIP :grin: 
```python
"TRIO302": "call to nursery.start/start_soon with resource from context manager opened on line {} something something nursery on line {}",
```
should likely add the variable names of both the nursery and the context manager too, and I don't currently point to the correct parameter when there's a multi-line call. (nor have a check for that)

(Also realizing now I don't cover `nursery.start(*bar)`, `nursery.start(**bar)`, passing bar several times, or passing stuff like `list(tuple(bar))`. But think it'll be pretty doable.)

**Comment by jakkdl** [comment]
Fixed stuff mentioned above, though code is still dirty and uncommented. Wanna take a pass through it tomorrow morning with a fresh head.

**Comment by jakkdl** [comment]
Merged in #34, so you should merge that before looking at the diff.

Code should be all pretty and commented now~ :sparkles:

File: flake8_trio.py [hunk 1]
```diff
@@ -55,6 +55,10 @@
         "`trio.[fail/move_on]_[after/at]` instead"
     ),
     "TRIO110": "`while <condition>: await trio.sleep()` should be replaced by a `trio.Event`.",
+    "TRIO111": (
+        "variable {2}, from context manager on line {0}, "
+        "passed to {3} from nursery opened on {1}, might get closed while in use"
```
Comment (Zac-HD) [inline flake8_trio.py:60]:
```suggestion
        "variable {} is usable within the context manager on line {}, but that "
        "will close before nursery opened on line {} - this is usually a bug.  "
        "Nurseries should generally be the inner-most context manager."
```

Not *super* happy with this, but I think it gets at the core problem pretty directly.  (message also requires some changes to the construction of the error, but dropping one and rearranging isn't too hard.)

**Review by Zac-HD — COMMENTED** [review]
Check looks great; some comments about the message (tricky!) and the tests (also tricky 😅) but I think we'll probably ship this today!

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,7 +2,8 @@
 *[CalVer, YY.month.patch](https://calver.org/)*
 
 ## Future
-- add TRIO112, nursery body with only a call to `nursery.start[_soon]` and not passing itself as a parameter can be replaced with a regular function call.
+- Add TRIO111: async context manager inside nursery. Nurseries should be outermost.
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply CHANGELOG.md:5]: oops, completely forgot to update this one after the error was redefined

File: tests/trio111.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        nursery.start(bar)  # safe
+with trio.open_process() as bar:
+    with trio.open_nursery() as nursery:
+        nursery.start(bar)  # safe
+with trio.open_process() as bar:
+    with trio.open_nursery() as nursery:
+        nursery.start(bar)  # safe
+with trio.open_process() as bar:
+    with trio.open_nursery() as nursery:
+        nursery.start(bar)  # safe
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply tests/trio111.py:62]: oops, artifact of moving tests out of function to be able to deindent them (and remove `async` in the process`

File: tests/trio111.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+    def foo_1():
+        nursery = noterror.something
+        with trio.open_process() as bar_2:
+            nursery.start(bar_2)  # safe
+
+    def foo_2():
+        nursery = noterror.something
+        with trio.open_process() as bar_2:
+            nursery.start(bar_2)  # safe
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply tests/trio111.py:75]: same as above, was an async with

File: flake8_trio.py (hunk 1 above)
Comment (jakkdl) [reply flake8_trio.py]: No need to drop/rearrange with indexed args, and redundant args getting ignored :P Long term maybe nice to do, but I think I'll leave the redundant arg for now for ease of putting it back.

There's also a question on how wordy one wants the error message to be, I personally prefer messages to be short enough to get shown as a one-liner in a status bar - esp the gist of them.

I'd maybe do something like
> variable {2}, from context manager on line {0}, passed to {3} from nursery opened outside of it on line {1}, might get invalidly accessed due to context manager closing before the nursery.
(though not sure about invalidly vs illegally vs ??)
But I'll leave yours as is and let you decide.

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,7 +2,8 @@
...
+- Add TRIO112: nursery body with only a call to `nursery.start[_soon]` and not passing itself as a parameter can be replaced with a regular function call.
```
[the remark this answers is not in the document]
Comment (jakkdl) [reply CHANGELOG.md]: Sounds good, also modified the error message.

**Review by Zac-HD — APPROVED** [review]
Spent a while trying to come up with a message I liked more... and couldn't.  Let's ship it!

**Pull request merged**